### PR TITLE
[WIP] Support repositories where scala project is located in sub directories

### LINKED ISSUE #46: Support nonstandard project layout.
Like the Play framework, which put the real sbt project in `/framework` folder.

### PR DESCRIPTION
**Pull request by exoego**
Closes #46 and #678. 

I am not sure the below design strategy is fine.
Need feedback from @fthomas, contributors and users.

I am considering to make this feature opt-in via an new entry in repository config, since traversing sub directories is I/O-heavy.
If the feature enabled, Scala Steward will traverse all subdirectories (with some exception like `target/` directories where no `build.sbt` exist) to find Scala projects.
I initially thought that users should specify sub projects directories by config.
But it could be human-error-prone if there are many sub projects added in case of monorepo.
So from ease-of-use perspective, all subdirectories should be traversed.

TODO:
- [ ] Add a fine-grain control of sub directories to be traversed (could be regex or array)

### CODE REVIEW

**Comment by robstoll** [comment]
I have a use case for this feature as well. I would suggest to add a config entry which expects a regex so that one can define in which paths it should search for subprojects. if one wants that all subdirectories are traversed, then one can simply provide `.*` but it would also allow for more fine-grained control

File: modules/core/src/main/scala/org/scalasteward/core/io/WorkspaceAlg.scala [hunk 1]
```diff
@@ -54,5 +56,16 @@ object WorkspaceAlg {
 
       override def repoDir(repo: Repo): F[File] =
         fileAlg.ensureExists(reposDir / repo.owner / repo.repo)
+
+      override def findSubProjectDirs(repo: Repo): F[List[File]] = {
+        def findSubProject(file: File): List[File] =
+          if (!file.isDirectory) List.empty
+          else if (file.contains(File("build.sbt"))) List(file)
```
Comment (robstoll) [inline modules/core/src/main/scala/org/scalasteward/core/io/WorkspaceAlg.scala]: I tried your commit locally, it does not work as you might think `File("build.sbt")` in WorkspaceAlg is relative to the execution path. Instead it should be something like: `file.path.resolve("build.sbt")`

Comment (exoego) [reply modules/core/src/main/scala/org/scalasteward/core/io/WorkspaceAlg.scala]: Thanks for testing and suggestion 👍
Updated based on your suggestion.

**Comment by robstoll** [comment]
@exoego any chance you rebase onto master and resolve conflicts? I already use your WIP :wink:

**Comment by robstoll** [comment]
Following the fix for the FIXME you left in the code
```
Index: modules/core/src/main/scala/org/scalasteward/core/nurture/NurtureAlg.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- modules/core/src/main/scala/org/scalasteward/core/nurture/NurtureAlg.scala	(revision fa159e36857af1588e2303af5b0326ebb486dbb0)
+++ modules/core/src/main/scala/org/scalasteward/core/nurture/NurtureAlg.scala	(revision 6c81527a844f438192d714550869248c7cf50b98)
@@ -79,7 +79,7 @@
       _ <- logger.info(s"Find updates for ${repo.show}")
       repoConfig <- repoConfigAlg.readRepoConfigOrDefault(repo)
       subProjects <- workspaceAlg.findSubProjectDirs(repo)
-      filtered <- findUpdatesInAllProjects(subProjects, repoConfig)
+      filtered <- findUpdatesInAllProjects(repo, subProjects, repoConfig)
       grouped = Update.group(filtered)
       _ <- logger.info(util.logger.showUpdates(grouped))
       baseSha1 <- gitAlg.latestSha1(repo, baseBranch)
@@ -105,13 +105,14 @@
     } yield result
 
   def findUpdatesInAllProjects(
+      repo: Repo,
       subProjectRoots: List[File],
       repoConfig: RepoConfig
   ): F[List[Update.Single]] =
     subProjectRoots
       .traverse { projectDir =>
         for {
-          updates <- sbtAlg.getUpdatesForRepo(projectDir)
+          updates <- sbtAlg.getUpdatesForRepo(repo, projectDir)
           filtered <- filterAlg.localFilterMany(repoConfig, updates)
         } yield filtered
       }
Index: modules/core/src/main/scala/org/scalasteward/core/sbt/SbtAlg.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- modules/core/src/main/scala/org/scalasteward/core/sbt/SbtAlg.scala	(revision fa159e36857af1588e2303af5b0326ebb486dbb0)
+++ modules/core/src/main/scala/org/scalasteward/core/sbt/SbtAlg.scala	(revision 6c81527a844f438192d714550869248c7cf50b98)
@@ -45,7 +45,7 @@
 
   def getUpdatesForProject(project: ArtificialProject): F[List[Update.Single]]
 
-  def getUpdatesForRepo(repoDir: File): F[List[Update.Single]]
+  def getUpdatesForRepo(repo: Repo, projectDir: File): F[List[Update.Single]]
 
   def runMigrations(repo: Repo, migrations: Nel[Migration]): F[Unit]
 }
@@ -130,16 +130,16 @@
           }
         } yield updates
 
-      override def getUpdatesForRepo(repoDir: File): F[List[Update.Single]] = {
+      override def getUpdatesForRepo(repo: Repo, projectDir: File): F[List[Update.Single]] = {
         val maybeClearCredentials = if (config.keepCredentials) Nil else List(setCredentialsToNil)
         val commands = maybeClearCredentials ++
           List(dependencyUpdates, reloadPlugins, dependencyUpdates)
         for {
+          repoDir <- workspaceAlg.repoDir(repo)
           updates <- withTemporarySbtDependency(repoDir) {
-            exec(sbtCmd(commands), repoDir).map(parser.parseSingleUpdates)
+            exec(sbtCmd(commands), projectDir).map(parser.parseSingleUpdates)
           }
-          // FIXME: Repos
-          originalDependencies <- cacheRepository.getDependencies(List.empty)
+          originalDependencies <- cacheRepository.getDependencies(List(repo))
           updatesUnderNewGroupId = originalDependencies.flatMap(
             UpdateService.findUpdateUnderNewGroup
           )
Index: modules/core/src/test/scala/org/scalasteward/core/sbt/SbtAlgTest.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- modules/core/src/test/scala/org/scalasteward/core/sbt/SbtAlgTest.scala	(revision fa159e36857af1588e2303af5b0326ebb486dbb0)
+++ modules/core/src/test/scala/org/scalasteward/core/sbt/SbtAlgTest.scala	(revision 6c81527a844f438192d714550869248c7cf50b98)
@@ -31,6 +31,7 @@
   }
 
   test("getUpdatesForRepo") {
+    val repo = Repo("fthomas", "refined")
     val repoDir = config.workspace / "fthomas/refined"
     val files = Map(
       repoDir / "project" / "build.properties" -> "sbt.version=1.2.6",
@@ -40,7 +41,7 @@
     files.foreach {
       case (file, content) =>
         val initialState = MockState.empty.copy(files = Map(file -> content))
-        val state = sbtAlg.getUpdatesForRepo(repoDir).runS(initialState).unsafeRunSync()
+        val state = sbtAlg.getUpdatesForRepo(repo, repoDir).runS(initialState).unsafeRunSync()
         state shouldBe MockState.empty.copy(
           files = Map(file -> content),
           commands = Vector(
@@ -68,9 +69,13 @@
   test("getUpdatesForRepo ignoring .jvmopts and .sbtopts files") {
     implicit val config: Config = MockContext.config.copy(ignoreOptsFiles = true)
     val sbtAlgKeepingCredentials = SbtAlg.create
+    val repo = Repo("fthomas", "refined")
     val repoDir = config.workspace / "fthomas/refined"
     val state =
-      sbtAlgKeepingCredentials.getUpdatesForRepo(repoDir).runS(MockState.empty).unsafeRunSync()
+      sbtAlgKeepingCredentials
+        .getUpdatesForRepo(repo, repoDir)
+        .runS(MockState.empty)
+        .unsafeRunSync()
 
     state shouldBe MockState.empty.copy(
       commands = Vector(
@@ -99,9 +104,13 @@
   test("getUpdatesForRepo keeping credentials") {
     implicit val config: Config = MockContext.config.copy(keepCredentials = true)
     val sbtAlgKeepingCredentials = SbtAlg.create
+    val repo = Repo("fthomas", "refined")
     val repoDir = config.workspace / "fthomas/refined"
     val state =
-      sbtAlgKeepingCredentials.getUpdatesForRepo(repoDir).runS(MockState.empty).unsafeRunSync()
+      sbtAlgKeepingCredentials
+        .getUpdatesForRepo(repo, repoDir)
+        .runS(MockState.empty)
+        .unsafeRunSync()
 
     state shouldBe MockState.empty.copy(
       commands = Vector(
```

**Comment by robstoll** [comment]
@fthomas I am willing to fix the conflicts and implement further fixes/features if required. Therefore my question, what needs to be done additionally that this can land in master?

**Comment by exoego** [comment]
Sorry, but closing since I am no longer working on this feature.
I hope someone take over.

**Pull request closed without merging**